### ISSUE TITLE
Pack in Visual Studio fails with transitive project reference in a clean state

### ISSUE THREAD

**Issue by TFTomSun** [dotnet/msbuild#5870]
Visual Studio Version: 16.8.0
Steps to reproduce:
- clone https://github.com/AppGates/AppGates.Build.Versioning
- pack external library
- clean referencing library
- pack referencing library 
-> build error, the transitive library is not built during pack

**Comment by BenVillalobos** [comment]
Team Triage: Does this fail only in Visual Studio? Have you tried on the command line?

**Comment by LeonsBuntis** [comment]
This fails also by directly using msbuild.

**Comment by TFTomSun** [comment]
In case somebody is interested. I implemented a workaround, by bypassing the visual studio pack and trigger a build with GeneratePackageOnBuild=true instead.

```xml
 <PropertyGroup>
    <IsPackableCopy>true</IsPackableCopy>
    <IsPackableCopy Condition="'$(IsPackable)' == 'false'">false</IsPackableCopy>
  </PropertyGroup>

  <!--Avoid the weird half dependency build of Visual Studio during pack-->
  <PropertyGroup >
        <IsPackable>false</IsPackable>
  </PropertyGroup>

  <Target Name="CustomPack" AfterTargets="Pack" >
    <PropertyGroup>
      <CustomPackBuildProperties>
        GeneratePackageOnBuild=true;
        NoBuild=false;
        BuildingProject=true;
        <!--BuildDependsOn=$(BuildDependsOn_Copy);-->
        <!--IsPackable=true;-->
        BuildingInsideVisualStudio=false;
        IsLocalBuild=$(IsLocalBuild);
      </CustomPackBuildProperties>
    </PropertyGroup>
    <Message Text="IsPackableCopy: $(IsPackableCopy)"/>
    <MSBuild
       Targets="Restore;Build"
      Condition="$(IsPackableCopy)"
             Projects="$(MSBuildProjectFullPath)"
             Properties="$(CustomPackBuildProperties)"/>
    
  </Target>



```

**Comment by BenVillalobos** [comment]
Team Triage: Out of curiosity, does this repro when you pass `/graph` as a parameter for your build?

**Comment by BenVillalobos** [comment]
@Forgind See if this repros.

**Comment by Forgind** [comment]
The problem here appears to be the intertwined dependencies of Pack and Build. I don't fully understand why, but it seems that in certain scenarios, Build depends on Pack, so there can't be an explicit dependency of Pack on Build because that would be a circular dependency in those cases. In this case, Pack is trying to execute the targets it depends on—which doesn't include Build—before Build executes. Specifically, it tries to execute GenerateNuspec, which expected Build to have executed already, but it hadn't, so this fails.

The easiest fix I found was to run with `/t:Build;Pack`. I'm still confused as to the logic [here](https://github.com/NuGet/NuGet.Client/pull/1148), but perhaps there's a solution buried there somewhere.

**Comment by TFTomSun** [comment]
@Forgind there's a build flag "GeneratePackageOnBuild". That might be the reason why build depends on pack.

**Comment by Forgind** [comment]
Maybe it makes sense to have Build sometimes depend on Pack from GeneratePackageOnBuild but Pack depend on Build's dependencies?

**Comment by TFTomSun** [comment]
@Forgind When I make some changes on the source code, I expect, that the source code is built before it is packed, right?
I've recently created a build package, that fixes the Visual Studio pack behavior to what I'd expect to happen, when I click on the pack context menu entry of a project file in VS:

- (Delta-)Build the root project and it's dependencies
- Pack the root project and it's dependencies

The package is [here](https://www.nuget.org/packages/AppGates.Net.Build.TransitivePack/) 
The project url is https://gitlab.com/appgates/net/build

**Comment by Forgind** [comment]
Correct. The tricky part is that the `Build` target doesn't really do anything—it's just something logic can hook into. It's intended to be the last target to actually be executed, after `BeforeBuild` and also after `AfterBuild`. That might be more confusing than necessary, but it's helpful in getting the final build result (succeeded, failed) to consider everything and not just the result of half of the total build process.

Your solution (and the workaround) should work, I think, but they both make the Build complete before Pack starts. I think that would mean that if there's a problem with Pack but not with Build, it will report success anyway.

Maybe a simpler solution to what I posed above that still retains the accurate reporting would be to have Pack depend on AfterBuild? Or hook into that some other way? I'll have to look more at build ordering before I know if that makes sense.

**Comment by TFTomSun** [comment]
@Forgind  I don't think that there's an issue with unexpected the success results. The package implementation is a bit different to what I posted above. Actually I turned of the complete default build and pack behavior in case Pack is called inside VisualStudio. There were simply to many weird side effects with the current VS behavior,especially with project dependencies and a mixture of single target and multi target projects.

Instead I call the build and the pack targets on my own in a target that I attached after Pack.
